### ISSUE TITLE
Add to save the highlighted/selected multiple tabs !

### ISSUE THREAD

**Issue by dnknn** [allen505/save-for-later-firefox#13]
> Save for later referred to as `SFL`.

Feature request: SFL can support to selected/highlighted multiple tabs with holding Ctrl/Shift !

I don't know if you know this browser basic feature ?

<details><summary><strong> Click to view: What is the selected/highlighted multiple tabs ❓</strong></summary>

![multi-selected-tabs](https://user-images.githubusercontent.com/10995302/108319576-92970480-71fc-11eb-9d39-bf291fdd9a16.gif)

1. You can try to holdding Ctrl/Shift/Ctrl+Shift on the tab-bar and click the non-current tab.
2. At this point, you have selected multiple tabs at the same time.
3. Finally, you can perform batch operations on these selected tabs. e.g. Ctrl+W to batch closes them, F5 to batch reload them ... can also drag them in batches. etc...

Example extension :  (In fact, many extensions support this feature)

- better-onetab
https://chrome.google.com/webstore/detail/eookhngofldnbnidjlbkeecljkfpmfpg
https://github.com/cnwangjie/better-onetab

- SingleFile
https://chrome.google.com/webstore/detail/mpiodijhokgodhhofbcjdecpffjipkle
http://github.com/gildas-lormeau/SingleFile

<img align=right width=40% src=https://user-images.githubusercontent.com/10995302/108318951-a0985580-71fb-11eb-949e-ba40fffe5aaf.png>

- TabCopy 
https://chrome.google.com/webstore/detail/micdllihgoppmejpecmkilggmaagfdmb

</details>

- So can you support for this browser kernel/API function (the 🔌 `highlighted` API can be used directly)

I don't know if it is this API-Doc :
https://developer.chrome.com/docs/extensions/reference/tabs/#event-onHighlighted

----

<img align=right width=30% src=https://user-images.githubusercontent.com/10995302/108317038-b0fb0100-71f8-11eb-8bbb-a2c199c6e28a.png>

- ## Current behavior:

　　SFL can only save all tabs in the window , which is very limited!
　　More often, users only want to save the selected/highlighted  tabs with holding Ctrl/Shift  (⚠In fact, the current tab is also selected/highlighted , it's just 1)

- ## Expected behavior:
　　When the multiple tabs be selected/highlighted with holding Ctrl/Shift/Ctrl+Shift,
At this time, it becomes an operation for the selected multiple tabs,
that is, at this time, it cannot remain as an operation for the current tab/page.

- **To sum up a sentence:**
SFL it can support for selected/highlighted multiple tabs with holding Ctrl/Shift/Ctrl+Shift.
---------------------------------------------------------------------

# 💡 Implementation plan : see 👇
![212_18150240](https://user-images.githubusercontent.com/10995302/108318189-5cf11c00-71fa-11eb-8efa-b375ec4260ce.png)

---

Whatever, must support this very smart/convenient/perfect feature, as for the UI/GUI is up to you ... ^_^

---

- ⚠ Of course, this function 👇 must also support the hotkey operating!

https://github.com/allen505/save-for-later/issues/12

---

💪👍 If can be achieved, will be perfect + very smart!

**Comment by allen505** [comment]
Hi @dnknn . Thanks for the suggestion.  
I wouldn't call this as a "_biggest problem_" with SFL😀 . This changes should be fairly easy to do. My idea for such a feature is to not add a new button, but just checking if there are tabs selected, and if it is, then only those selected tabs they will be saved, silently.  

To implement this feature we need to use the `highlighted` condition in [this line](https://github.com/allen505/save-for-later/blob/master/popup/main.js#L73).  
Reference: https://developer.mozilla.org/en-US/docs/Mozilla/Add-ons/WebExtensions/API/tabs/Tab

**Comment by dnknn** [comment]
> My idea for such a feature is to not add a new button, but just checking if there are tabs selected, and if it is, then only those selected tabs they will be saved, silently.

💪👍 Yes, Great idea! Be smarter and keep lightweight!


> To implement this feature we need to use the `highlighted` condition in [this line](https://github.com/allen505/save-for-later/blob/master/popup/main.js?rgh-link-date=2021-02-19T16%3A03%3A26Z#L73).

I don't understand the code 🤦

**Comment by dnknn** [comment]
@allen505 
> My idea for such a feature is to not add a new button, but just checking if there are tabs selected, and if it is, then only those selected tabs they will be saved, silently.


- But here comes the question, what if the user only wants to save the current 1 tab ?

So you should add options like this: ⬜ If there are no `highlighted` tabs, the default behavior is to save the current **window** ? or the current **tab** ? , this option can perfectly solve this problem!

---

If follow your current point of view, then the following :

- ## 💥 **Current behavior:**
　　　Save the current window or `highlighted` tabs, but **cannot** save the current 1 tab, This still means, unwanted tabs will be forced to be saved (If you don’t select ⩾2 tabs)


- ## ❥ **Expected Default/Optional behavior:**
　　　Save the current 1 tab or `highlighted`/selected multiple tabs.

**Comment by allen505** [comment]
> I don't understand the code  

I added it for a future reference when anyone works on this. I should've made a separate comment.  

You bring up a valid issue. So my thinking is that saving a single tab should not be possible at all, not even as a settings option. The reasoning behind it is that saving a single tab is done by using Bookmarks. This extension should be used to save multiple tabs or an entire window.

**Comment by dnknn** [comment]
> saving a single tab should not be possible at all

> The reasoning behind it is that saving a single tab is done by using Bookmarks.

> This extension should be used to save multiple tabs or an entire window.


Different usage scenarios, The positioning of this extension is to read later , It means to delete it after reading it, and bookmarks are similar to permanent storage.

If you want to save multiple tabs, there are too many ways to implement it even without this extension, and you can even copy the URLs of multiple tabs and save them elsewhere, so I hope you think about it.

In other words, users can use this extension as a **temporarily** saved scene, temporarily saved, or called **temporary** storage first

**Comment by dnknn** [comment]
@allen505  I think a single tab save is also  save!
----
There are also many users who do not display the bookmark-bar (`Ctrl+Shift+B`),
Think about it: Will you save a temporary single tab as a bookmark? Come over and read it like delete it?

The key issue is that bookmarks are administrative in nature, and the extension should be positioned as casual and temporary. Of course, personal opinion...you decide ^_^

----
In addition, the main reason I chose the extension is that the page is not closed when saving, And lightweight / miniaturization , Very much in line with the name of the extension "Save for later " , i.e. Fast/easy to 💾save, and Fast/easy to 🗑delete.

When need to close the page after save, I will switch to [better-onetab](https://chrome.google.com/webstore/detail/eookhngofldnbnidjlbkeecljkfpmfpg) extension (Can save single or multiple) or  [Panic Button](https://chrome.google.com/webstore/detail/faminaibgiklngmfpfbhmokfmnglamcm) extension. 

So the positioning/function of each extension is different.

**Comment by allen505** [comment]
I still feel that highlight and save for a single tab is not necessary and that it is a rare occurrence. On such occasions you can separate the single tab into a separate window by pulling it out of the current window and saving the newly created window. This way the single tab can be easily created.

**Comment by allen505** [comment]
I have implemented this feature in my latest commit. If you wish to test the new version locally please follow [these instructions](https://github.com/allen505/save-for-later#development-setup).

**Comment by dnknn** [comment]
> I have implemented this feature in my latest commit

Thank you very much!!!

The test results are very good, in line with expectations, and enough! The single tab is up to you !
